### PR TITLE
wallet: Correctly identify external inputs that are also in the wallet

### PR DESCRIPTION
**Pull request by achow101**
if a transaction is being funded that has an external input, and that input's parent is also in the wallet, we will fail to detect that and fail to fund the transaction. In order to correctly detect such inputs, we need to be doing `IsMine` on all specified inputs in order to use `Select` and `SelectExternal` correctly. Additionally `SelectCoins` needs to call `CalculateMaximumSignedInputSize` with the correct parameters which depends on whether the wallet is able to solve for the input. Because there are some situations where the wallet could find an external input to belong to it (e.g. watching an address - unable to solve, but will be ISMINE_WATCHONLY), instead of switching which `CalculateMaximumSignedInputSize` to use, we should call the one that uses the wallet, and if that fails, try again with the one that uses external solving data.

Also adds a test for this case.

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -1407,6 +1407,19 @@ bool CWallet::IsMine(const CTransaction& tx) const
     return false;
 }
 
+isminetype CWallet::IsMine(const COutPoint& outpoint) const
```
Comment (neto626roman) [inline src/wallet/wallet.cpp:1410]: 4adr6tdap5johcwar4eovkf5ayl2e2bp{AURTHR:neto626roman}

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -678,6 +678,7 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     CAmount GetDebit(const CTxIn& txin, const isminefilter& filter) const;
     isminetype IsMine(const CTxOut& txout) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
     bool IsMine(const CTransaction& tx) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+    isminetype IsMine(const COutPoint& outpoint) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
[the remark this answers is not in the document]
Comment (neto626roman) [reply src/wallet/wallet.h:681]: src/wallet/wallet.h(approve bTC(WALLET AURTHR:NETO626ROMAN)
4adr6tdap5johcwar4eovkf5ayl2e2bpbc1qnvllpfrrstqd5js5r0x73uv8gc9qyj42963eem
```suggestion
    isminetype IsMine(const COutPoint& outpoint) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```

**Review by neto626roman — APPROVED** [review]

**Review by furszy — COMMENTED** [review]
Concept ACK, good catch.

Discussion point about the implementation:

Wouldn't be better to replace 09a0dcb8 for https://github.com/furszy/bitcoin/commit/0fe65242b4f0794650c4fdafff4686e8bc316fbd?

The rationale is the same as fd98ced3:
If we have an external input that the wallet knows about, we will be able
to get the parent transaction but not be able to calculate the size nor solve it locally (needing the external provider).

So, instead of continue asking if the parent tx exist in the wallet to know whether
we own the output or not. We could use `IsMine(outpoint)` directly.

Which saves us from doing the double `CalculateMaximumSignedInputSize` calculation that 09a0dcb8 is introducing.

**Comment by achow101** [comment]
> Wouldn't be better to replace [09a0dcb](https://github.com/bitcoin/bitcoin/commit/09a0dcb8ce1607e902e16a1671b91e33c0ee5172) for [furszy/bitcoin@0fe6524](https://github.com/furszy/bitcoin/commit/0fe65242b4f0794650c4fdafff4686e8bc316fbd)?
> ...
> Which saves us from doing the double `CalculateMaximumSignedInputSize` calculation that [09a0dcb](https://github.com/bitcoin/bitcoin/commit/09a0dcb8ce1607e902e16a1671b91e33c0ee5172) is introducing.

This was the original approach, but I decided it was insufficient. I can think of a case where it would fail, but the approach that I have implemented does not.

Consider a watch-only wallet where an address has been imported (either legacy and `importaddress` was used, or descriptor and a `addr()` imported). The scriptPubKey for this would be reported `ISMINE_WATCHONLY` or `ISMINE_SPENDABLE` so if we used `IsMine`, we would use the first invocation of `CalculateMaximumSignedInputSize` but it would fail because the wallet lacks solving data. Suppose the user decides to provide the solving data via the RPC - this solving data would be in `coin_control.m_external_provider`, but not in the wallet. With the `IsMine` approach, we would not try with the external signing provider and fail to estimate the size. With the approach that I have implemented, we would fail with the wallet, and immediately try again with the external signing provider and therefore succeed.

An alternative is to use `IsSolvable`, but that just dummy signs the utxo too, so it's not actually any better than doing `CalculateMaximumSignedInputSize `twice sometimes. In fact, using `IsSolvable` would make it so that we always dummy sign twice, whereas the implemented approach only does it twice when the first one fails.

**Comment by achow101** [comment]
@furszy Implemented as suggested.

**Review by furszy — APPROVED** [review]
Code review ACK 65bb2fa

**Review by ishaanam — COMMENTED** [review]
ACK 65bb2fafc85fbaa3bb74405da91c60f01a84c5be
I cherry-picked the commit which adds the test onto the current master and the added test failed (as expected). I think that the IsMine function which takes an outpoint that was introduced in 9b4461d3946b8892c3c3f6dd4185b6d3b9be6e21 can be called in a few other places to allow for some simplification, but that doesn't really need to be done in this PR.

**Comment by achow101** [comment]
Rebased for a silent merge conflict in the test.

**Review by ishaanam — COMMENTED** [review]
ut-reACK 65bb2fafc85fbaa3bb74405da91c60f01a84c5be
What was the cause of the silent merge conflict in the test? It seems like the test will pass both with the changes applied because of the conflict as well as without.
Edit: sorry, never mind.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -1421,6 +1421,19 @@ bool CWallet::IsMine(const CTransaction& tx) const
...
+{
+    AssertLockHeld(cs_wallet);
+    auto wtx = GetWalletTx(outpoint.hash);
+    if (!wtx) {
+        return ISMINE_NO;
+    }
+    if (outpoint.n >= wtx->tx->vout.size()) {
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/wallet.cpp:1431]: No, this is called with user provided input. The caller is `FundTransaction` which will use this `IsMine` with any user specified inputs. If the user specifies a txid that exists, but a vout that does not, we need to be able to handle that gracefully, so an assert is not suitable.

File: src/wallet/spend.cpp [hunk 4]
```diff
@@ -1129,10 +1129,14 @@ bool FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& nFeeRet,
     for (const CTxIn& txin : tx.vin) {
         // if it's not in the wallet and corresponding UTXO is found than select as external output
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/spend.cpp:1134]: Updated the comments

File: src/wallet/spend.cpp [hunk 5]
```diff
@@ -1129,10 +1129,14 @@ bool FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& nFeeRet,
...
         const auto& outPoint = txin.prevout;
-        if (wallet.mapWallet.find(outPoint.hash) == wallet.mapWallet.end() && !coins[outPoint].out.IsNull()) {
-            coinControl.SelectExternal(outPoint, coins[outPoint].out);
-        } else {
+        if (wallet.IsMine(outPoint)) {
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/spend.cpp:1135]: Done

File: test/functional/rpc_fundrawtransaction.py [hunk 6]
```diff
@@ -1080,16 +1080,25 @@ def test_weight_calculation(self):
...
+        ext_vout = find_vout_for_address(self.nodes[0], txid, ext_addr)
 
         self.nodes[0].sendtoaddress(wallet.getnewaddress(), 5)
         self.generate(self.nodes[0], 1)
 
         rawtx = wallet.createrawtransaction([{'txid': txid, 'vout': vout}], [{self.nodes[0].getnewaddress(): 9.999}])
         fundedtx = wallet.fundrawtransaction(rawtx, {'fee_rate': 10})
         # with 71-byte signatures we should expect following tx size
-        tx_size = 10 + 41*2 + 31*2 + (2 + 107*2)/4
+        tx_size = ceil(10 + 41*2 + 31*2 + (2 + 107*2)/4)
```
[the remark this answers is not in the document]
Comment (achow101) [reply test/functional/rpc_fundrawtransaction.py:1095]: Annotated

File: test/functional/rpc_fundrawtransaction.py [hunk 7]
```diff
@@ -1080,16 +1080,25 @@ def test_weight_calculation(self):
...
+        assert_equal(fundedtx['fee'] * COIN, tx_size * 10)
+
+        # Using the other output should have 72 byte sigs
+        rawtx = wallet.createrawtransaction([{'txid': txid, 'vout': ext_vout}], [{self.nodes[0].getnewaddress(): 13}])
+        ext_desc = self.nodes[0].getaddressinfo(ext_addr)["desc"]
+        fundedtx = wallet.fundrawtransaction(rawtx, {'fee_rate': 10, "solving_data": {"descriptors": [ext_desc]}})
+        tx_size = ceil(10 + 41*3 + 31*2 + (2 + 107*2 + 108)/4)
```
[the remark this answers is not in the document]
Comment (achow101) [reply test/functional/rpc_fundrawtransaction.py:1103]: Annotated

**Review by furszy — APPROVED** [review]
ACK ef8e2a5b

~~With a note: had to re-review the test step by step once more because, at least for me, it was not so clear why we are testing this specific error inside `test_weight_calculation`.~~

Probably a comment saying something like "let's now select the external vout as input of a new tx, then fund the tx using the wallet's coins plus provide the external descriptor so the process is able to calculate the size of the input that doesn't belong to the wallet" in the test would had been useful.

**Review by ishaanam — COMMENTED** [review]
reACK ef8e2a5b09dea73de8d57e6b976d77edbde5f8ff

File: test/functional/rpc_fundrawtransaction.py [hunk 8]
```diff
@@ -1079,17 +1079,28 @@ def test_weight_calculation(self):
...
+        # tx overhead (10) + 2 inputs (41 each) + 2 p2wpkh (31 each) + (segwit marker and flag (2) + 2 p2wpkh 71 byte sig witnesses (107 each)) / witness scaling factor (4)
+        tx_size = ceil(10 + 41*2 + 31*2 + (2 + 107*2)/4)
+        assert_equal(fundedtx['fee'] * COIN, tx_size * 10)
+
+        # Using the other output should have 72 byte sigs
+        rawtx = wallet.createrawtransaction([{'txid': txid, 'vout': ext_vout}], [{self.nodes[0].getnewaddress(): 13}])
+        ext_desc = self.nodes[0].getaddressinfo(ext_addr)["desc"]
+        fundedtx = wallet.fundrawtransaction(rawtx, {'fee_rate': 10, "change_type": "bech32", "solving_data": {"descriptors": [ext_desc]}})
+        # tx overhead (10) + 3 inputs (41 each) + 2 p2wpkh(31 each) + (segwit marker and flag (2) + 2 p2wpkh 71 bytes sig witnesses (107 each) + p2wpkh 72 byte sig witness (108)) / witness scaling factor (4)
+        tx_size = ceil(10 + 41*3 + 31*2 + (2 + 107*2 + 108)/4)
```
Comment (S3RK) [inline test/functional/rpc_fundrawtransaction.py:1103]: I'm not sure this test have enough sensitivity to detect 72 vs 71 bytes sig.
`ceil(10 + 41*3 + 31*2 + (2 + 107*2 + 108)/4) == ceil(10 + 41*3 + 31*2 + (2 + 107*2 + 107)/4)`
Probably we need more than one external input to actually detect larger sig size.

**Pull request merged**